### PR TITLE
Check that sample and ancestor files are finalised

### LINKED ISSUE #125: Check if samples and ancestors files are finalised on input

### PR DESCRIPTION
**Pull request by hyanwong**
Fixes #125 - as raised by @brianzhang01

### CODE REVIEW

File: tsinfer/formats.py [hunk 1]
```diff
@@ -435,6 +435,13 @@ def _check_write_modes(self):
         if self._mode not in (self.EDIT_MODE, self.BUILD_MODE):
             raise ValueError("Invalid operation: must be in edit or build mode")
 
+    def _check_finalised(self):
+        if not self.finalised:
+            error_msg = "The {} file".format(self.format_name)
```
Comment (brianzhang01) [inline tsinfer/formats.py]: I'm pretty sure we should change `"The {} file"` --> `"The {} file "`, with a space after "file"

Comment (hyanwong) [reply tsinfer/formats.py]: Good point

File: tests/test_inference.py [hunk 2]
```diff
@@ -54,6 +55,77 @@ def assertTreeSequencesEqual(self, ts1, ts2):
         self.assertEqual(t1.mutations, t2.mutations)
 
 
+class TestUnfinalisedErrors(unittest.TestCase):
+    def test_generate_ancestors(self):
+        with tempfile.NamedTemporaryFile() as tmpfile:
+            for path in [None, tmpfile.name]:
+                sample_data = tsinfer.SampleData(path=path, sequence_length=2)
+                sample_data.add_site(1, genotypes=[0, 1, 1, 0], alleles=["G", "C"])
+                print(dir(sample_data))
```
Comment (jeromekelleher) [inline tests/test_inference.py]: Spamming the output here.

File: tests/test_inference.py [hunk 3]
```diff
@@ -54,6 +55,77 @@ def assertTreeSequencesEqual(self, ts1, ts2):
...
+                # match_ancestors fails when samples unfinalised
+                unfinalised = tsinfer.SampleData(sequence_length=2)
+                unfinalised.add_site(1, genotypes=[0, 1, 1, 0], alleles=["G", "C"])
+                self.assertRaises(
+                    ValueError, tsinfer.match_ancestors, unfinalised, anc)
+
+    def test_match_ancestors_unfinalised_ancestors(self):
+        with tempfile.NamedTemporaryFile() as tmpfile:
+            for path in [None, tmpfile.name]:
+                sample_data = tsinfer.SampleData(sequence_length=2)
```
Comment (jeromekelleher) [inline tests/test_inference.py]:
```python
with  tsinfer.SampleData(sequence_length=2) as sample_data:
     sample_data.add_site(1, genotypes=[0, 1, 1, 0], alleles=["G", "C"])
```
would be preferable for the correct case I think.

File: tests/test_inference.py [hunk 4]
```diff
@@ -54,6 +55,77 @@ def assertTreeSequencesEqual(self, ts1, ts2):
...
+
+    def test_make_ancestor_data(self):
+        with tempfile.NamedTemporaryFile() as tmpfile:
+            for path in [None, tmpfile.name]:
+                sample_data = tsinfer.SampleData(path=path, sequence_length=2)
+                sample_data.add_site(1, genotypes=[0, 1, 1, 0], alleles=["G", "C"])
+                self.assertRaises(ValueError, tsinfer.AncestorData, sample_data)
+
+    def test_match_ancestors_unfinalised_samples(self):
+        with tempfile.NamedTemporaryFile() as tmpfile:
```
Comment (jeromekelleher) [inline tests/test_inference.py]: I don't think we need to repeat the check on "with a file or not" for each test. We can probably just do this once each for samples files and ancestors files.

File: tests/test_inference.py [hunk 5]
```diff
@@ -54,6 +55,77 @@ def assertTreeSequencesEqual(self, ts1, ts2):
...
+                sample_data = tsinfer.SampleData(path=path, sequence_length=2)
+                sample_data.add_site(1, genotypes=[0, 1, 1, 0], alleles=["G", "C"])
+                print(dir(sample_data))
+                self.assertRaises(ValueError, tsinfer.generate_ancestors, sample_data)
+                sample_data.finalise()
+                tsinfer.generate_ancestors(sample_data)
+
+    def test_make_ancestor_data(self):
+        with tempfile.NamedTemporaryFile() as tmpfile:
+            for path in [None, tmpfile.name]:
```
Comment (jeromekelleher) [inline tests/test_inference.py]: Annoyingly, this won't work on Windows. What you have to do is 
```python
with tempfile.TemporaryDirectory() as tmpdir:
    tmpfile = pathlib.Path(tmpdir) / "tmp.samples"
    # and so on
```

File: tsinfer/formats.py [hunk 6]
```diff
@@ -435,6 +435,13 @@ def _check_write_modes(self):
...
+            if self.path:
```
Comment (jeromekelleher) [inline tsinfer/formats.py]:
```python
if self.path is not None:
```

File: tests/test_inference.py (hunk 2 above)
Comment (hyanwong) [reply tests/test_inference.py]: Well spotted. Ta.

File: tests/test_inference.py (hunk 5 above)
Comment (hyanwong) [reply tests/test_inference.py]: It looks from `test_formats.py` that even using TemporaryDirectory these are skipped in windows anyway, so I'll just skip.

File: tests/test_inference.py (hunk 4 above)
Comment (hyanwong) [reply tests/test_inference.py]: OK

**Comment by jeromekelleher** [comment]
Looks good thanks, merging.

**Pull request merged**